### PR TITLE
Optimizing NonMatching::create_coupling_sparsity_pattern 

### PR DESCRIPTION
**Pull request by GivAlz**
The current implementation of the two functions to create the coupling matrix make a huge number of calls to compute_point_locations.
With this PR I unify these calls in a single one and reduce the number of steps used in the for cycles.

If anyone has suggestions I am here; I also have a question: 
say "cell" is a cell iterator to the first active cell of the triangulation. If I want to go to the "n-th cell after the first cell" is there a better way than:
```
for(unsigned int i=0; i<n;++i)
 ++cell;
```
As far as I can see nothing like "cell += n" is working.
Currently I am using two while cycles to do this sort of change: it works, but to me it feels like something is "off".

Thank you!
Best,
Giovanni

### CODE REVIEW

File: source/non_matching/coupling.cc [hunk 1]
```diff
@@ -140,6 +136,25 @@ namespace NonMatching
       if (immersed_c[i])
         immersed_gtl[i] = j++;
 
+    const unsigned int           n_q_points = quad.size();
+    std::vector<Point<spacedim>> AllPoints(
+      immersed_dh.get_triangulation().n_active_cells() * n_q_points);
+    {
+      unsigned int c = 0;
+      for (auto cell : immersed_dh.active_cell_iterators())
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]:
```suggestion
      for (const auto &cell : immersed_dh.active_cell_iterators())
```

**Comment by masterleinad** [comment]
> I also have a question:
> say "cell" is a cell iterator to the first active cell of the triangulation. If I want to go to the "n-th cell after the first cell" is there a better way than:
> 
> ```
> for(unsigned int i=0; i<n;++i)
>  ++cell;
> ```
> 
> As far as I can see nothing like "cell += n" is working.
> Currently I am using two while cycles to do this sort of change: it works, but to me it feels like something is "off".

Currently, there is no such functionality. Since you can't rely on the order of cells, it is unusual to increment or decrement a cell iterator by more than one element. You can use [`std::advance`](https://en.cppreference.com/w/cpp/iterator/advance) instead.

File: source/non_matching/coupling.cc (hunk 1 above)
Comment (GivAlz) [reply source/non_matching/coupling.cc]: Oops...thank you for noticing it!

**Comment by GivAlz** [comment]
The fact that I can't rely on cell order is troubling...in this case I am running twice the same iterator, without changing the mesh or anything...can I rely on its order?

**Comment by masterleinad** [comment]
What I meant was that the order of cells is an implementation detail. You should not assume anything about the successor or ancestor of a given cell. On the other hand, the ordering is fixed as long as you don't change the mesh.

**Comment by GivAlz** [comment]
Thank you @masterleinad for that std::advance tip, now the code looks much cleaner!

**Comment by GivAlz** [comment]
...and sorry, I just realized this algorithm would be extremely difficult to implement for the assemble matrix case...so I rewrote it: this one is much cleaner and I believe better.

**Comment by GivAlz** [comment]
@luca-heltai  as soon as you're okay with this second version I shall squash commits and force-push them

File: source/non_matching/coupling.cc [hunk 2]
```diff
@@ -140,6 +136,26 @@ namespace NonMatching
       if (immersed_c[i])
         immersed_gtl[i] = j++;
 
+    const unsigned int n_q_points = quad.size();
+    const unsigned int n_active_c =
+      immersed_dh.get_triangulation().n_active_cells();
+    std::vector<Point<spacedim>> AllPoints(n_active_c * n_q_points);
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: Usually, we don't capitalize variable names.

File: source/non_matching/coupling.cc [hunk 3]
```diff
@@ -160,23 +176,51 @@ namespace NonMatching
...
+    for (unsigned int i = 0; i < maps.size(); ++i)
       {
-        // Reinitialize the cell and the fe_values
-        fe_v.reinit(cell);
-        cell->get_dof_indices(dofs);
+        // Quadrature points should be reasonably clustered:
+        // the following index keeps track of the last id
+        // where the current cell was inserted
+        unsigned int last_id = std::numeric_limits<unsigned int>::max();
+        for (const auto &idx : maps[i])
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: This is an `unsigned int` if I am not mistaken and we would typically create a copy.

File: source/non_matching/coupling.cc [hunk 4]
```diff
@@ -160,23 +176,51 @@ namespace NonMatching
...
-        const auto  cpm   = GridTools::compute_point_locations(cache, Xpoints);
-        const auto &cells = std::get<0>(cpm);
+    for (unsigned int i = 0; cell != endc; ++cell, ++i)
+      {
+        // Reinitialize the cell
+        cell->get_dof_indices(dofs);
 
-        for (unsigned int c = 0; c < cells.size(); ++c)
+        // List of outer cells
+        const auto &cells = cell_sets[i];
```
Comment (masterleinad) [inline source/non_matching/coupling.cc]: This is the only place where `i` is used. Hence, you could initialize `i` outside the loop, replace this line by
```
 const auto &cells = cell_sets[i];
```
and switch to a range-based for loop if you prefer that.

Comment (GivAlz) [reply source/non_matching/coupling.cc]: Ok, I guess you forgot the ++ inside cell_sets, now I have:
`const auto &cells = cell_sets[i++];`
Thank you!

Comment (masterleinad) [reply source/non_matching/coupling.cc]: Yes, of course.

File: source/non_matching/coupling.cc [hunk 5]
```diff
@@ -140,6 +136,26 @@ namespace NonMatching
...
+    const unsigned int n_q_points = quad.size();
+    const unsigned int n_active_c =
+      immersed_dh.get_triangulation().n_active_cells();
+    std::vector<Point<spacedim>> all_points(n_active_c * n_q_points);
+    {
+      unsigned int c = 0;
+      for (const auto &cell : immersed_dh.active_cell_iterators())
+        {
+          // Reinitialize the cell and the fe_values
+          fe_v.reinit(cell);
```
Comment (bangerth) [inline source/non_matching/coupling.cc]: I believe this is the only place where you use `fe_v` in this function. Move its declaration to line 144, inside the current `{...}` block.

File: source/non_matching/coupling.cc [hunk 6]
```diff
@@ -160,23 +176,49 @@ namespace NonMatching
...
+    for (const auto &cell : immersed_dh.active_cell_iterators())
+      {
+        // Reinitialize the cell
+        cell->get_dof_indices(dofs);
 
-        // Get a list of outer cells, qpoints and maps.
-        const auto  cpm   = GridTools::compute_point_locations(cache, Xpoints);
-        const auto &cells = std::get<0>(cpm);
+        // List of outer cells
+        const auto &cells = cell_sets[i++];
```
Comment (bangerth) [inline source/non_matching/coupling.cc]: These sorts of statements are really difficult to read. You're creating an alias for `cell_sets[i]`, but then you also *as a side effect* increment `i`. It would be much easier to read this kind of code if you just write `const auto &cells = cell_sets[i]` and increment `i` elsewhere -- the best place would probably be the end of the loop block.

(While there, maybe also call the increment variable `cell_number` or some such to make it clearer what it corresponds to.)

File: source/non_matching/coupling.cc (hunk 5 above)
Comment (GivAlz) [reply source/non_matching/coupling.cc]: Wow, I would have never though about that. Thanks!

**Comment by GivAlz** [comment]
Ok, thank you @bangerth for your last suggestions. Just pushed the new version.

File: source/non_matching/coupling.cc (hunk 5 above)
Comment (bangerth) [reply source/non_matching/coupling.cc]: I'll even tell you why I suggested this. When I looked at the code, I was wondering what the update flags for the FEValues object were because you were only looking at the quadrature point locations. So I had to scroll up to see where that object was declared. In other words, for a reader of a code, it's *convenient* to have the declaration as far down in a function as possible because then the declaration/initialization and the use of a variable are as close as possible together. 

Secondly, it's useful to put a declaration in the inner-most scope possible so that the variable is destroyed as soon as possible -- so that (i) the memory used up is released again, and (ii) if a reader wonders whether the variable is used elsewhere, one doesn't have to scroll down very far.

In summary, it's generally a good idea to limit the lifetime of local variables as much as possible by pushing declarations down in the function and into the innermost scope possible.
